sip/sip-002-smart-contract-language.md

```diff
@@ -0,0 +1,493 @@
+_fetch-entry_ will always return an object whose size is equal
+to the specified value type of the map.
+
+A complete proof for the static runtime analysis of smart contracts
+will be included with the implementation of the language.
+
+# Deploying the Smart Contract
+
+Smart contracts on the Stacks blockchain will be deployed directly as
+source code. The goal of the smart contracting language is that the
+code of the contract defines the _ground truth_ about the intended
+functionality of the contract. While seemingly banal, many systems
+chose instead to use a compiler to translate from a friendly
+high-level language to a lower-level language deployed on the
+blockchain. Such an architecture is needlessly dangerous. A bug in
+such a compiler could lead to a bug in a deployed smart contract when
+no such bug exists in the original source. This is problematic for
+recovery --- a hard fork to "undo" any should-have-been invalid
+transactions would be contentious and potentially create a rift in the
+community, especially as it will not be easy to deduce which contracts
+exactly were affected and for how long. In contrast, bugs in the VM
+itself present a more clear case for a hard fork: the smart contract
+was defined correctly, as everyone can see directly on the chain, but
+illegal transactions were incorrectly marked as valid.
+
+# Virtual Machine API
+
+From the perspective of other components of `blockstack-core`, the
+smart contracting VM will provide the following interface:
+
+```
+connect-to-database(db)
+
+publish-contract(
+  contract-source-code)
+
+  returns: contract-identifier
+
+execute-contract-transaction(
+  contract-identifier,
+  transaction-name,
+  sender-principal,
+  transaction-arguments)
+
+  returns: true or false if the transaction executed successfully
+```
+
+
+### Database Requirements and Transaction Accounting
+
+The smart contract VM needs to interact with a database somewhat
+directly: the effects of an `insert-entry!` or `set-entry!` call are
+realized later in the execution of the same transaction. The database
+will need to support fairly fine-grained rollbacks as some contract
+calls within a transaction's execution may fail, triggering a
+rollback, while the transaction execution continues and successfully
+completes other database operations.
+
+The database API provided to the smart contract VM, therefore, must be
+capable of (1) quickly responding to `fetch-entry` queries, which are
+essentially simply key-value _gets_ on the materialized view of the
+operation log. The operation log itself is simply a log of the
+`insert-entry!` and `set-entry!` calls. In addition to these
+operations, the smart contract VM will be making token transfer calls.
+The databasse log should track those operations as well.
+
+In order to aid in accounting for the database operations created by
+a given transaction, the underlying database should store, with each
+operation entry, the corresponding transaction identifier.
```

Comment:
This will be expanded in a future SIP to say that the database must store enough information to reconstruct each block, so they can be relayed to bootstrapping peers.

Reply:
Added